kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ProxyReconciler.java

```diff
@@ -5,28 +5,192 @@
+import io.javaoperatorsdk.operator.api.reconciler.ErrorStatusHandler;
+import io.javaoperatorsdk.operator.api.reconciler.ErrorStatusUpdateControl;
 import io.javaoperatorsdk.operator.api.reconciler.Reconciler;
 import io.javaoperatorsdk.operator.api.reconciler.UpdateControl;
 import io.javaoperatorsdk.operator.api.reconciler.dependent.Dependent;
 
 import io.kroxylicious.kubernetes.api.v1alpha1.KafkaProxy;
 import io.kroxylicious.kubernetes.api.v1alpha1.KafkaProxyBuilder;
+import io.kroxylicious.kubernetes.api.v1alpha1.kafkaproxystatus.Conditions;
+import io.kroxylicious.kubernetes.api.v1alpha1.kafkaproxystatus.ConditionsBuilder;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
 
+// @formatter:off
 @ControllerConfiguration(dependents = {
-        @Dependent(type = ProxyConfigSecret.class),
-        @Dependent(type = ProxyDeployment.class),
-        @Dependent(type = ProxyService.class)
+        @Dependent(
+                name = ProxyReconciler.CONFIG_DEP,
+                type = ProxyConfigSecret.class),
+        @Dependent(
+                name = ProxyReconciler.DEPLOYMENT_DEP,
+                type = ProxyDeployment.class,
+                dependsOn = { ProxyReconciler.CONFIG_DEP },
+                readyPostcondition = DeploymentReadyCondition.class
+        ),
+        @Dependent(
+                name = ProxyReconciler.CLUSTERS_DEP,
+                type = ClusterService.class,
+                dependsOn = { ProxyReconciler.DEPLOYMENT_DEP }
+                //useEventSourceWithName = ProxyReconciler.CLUSTERS_DEP
+        )
 })
-public class ProxyReconciler implements Reconciler<KafkaProxy> {
+// @formatter:on
+public class ProxyReconciler implements
+        Reconciler<KafkaProxy>,
+        ErrorStatusHandler<KafkaProxy> {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProxyReconciler.class);
+
+    public static final String CONFIG_DEP = "config";
+    public static final String DEPLOYMENT_DEP = "deployment";
+    public static final String CLUSTERS_DEP = "clusters";
+
+    @Override
     public UpdateControl<KafkaProxy> reconcile(KafkaProxy primary,
                                                Context<KafkaProxy> context) {
         return UpdateControl.patchStatus(
-                new KafkaProxyBuilder(primary).editOrNewStatus()
-                        .withObservedGeneration(primary.getMetadata().getGeneration())
-                        .endStatus()
-                        .build());
+                buildStatus(primary, null));
+    }
+
+    @Override
+    public ErrorStatusUpdateControl<KafkaProxy> updateErrorStatus(KafkaProxy primary,
+                                                                  Context<KafkaProxy> context,
+                                                                  Exception exception) {
+        // Post-condition: status.conditions should be in a canonical order (to avoid non-terminating reconciliations)
```

Comment:
Do we/should we have tests for the post conditions?
What is the canonical order?

Reply:
We only have a single condition currently, so an assertion on ordering is pointless. I think the testing in `ProxyReconcilerTest` is sufficient for now.